src/diffusers/models/transformer_2d.py

```diff
@@ -289,15 +291,28 @@ def forward(
 
         # 2. Blocks
         for block in self.transformer_blocks:
-            hidden_states = block(
-                hidden_states,
-                attention_mask=attention_mask,
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_attention_mask,
-                timestep=timestep,
-                cross_attention_kwargs=cross_attention_kwargs,
-                class_labels=class_labels,
-            )
+            if self.training and self.gradient_checkpointing:
+                hidden_states = torch.utils.checkpoint.checkpoint(
```

Comment:
SDXL has 10 Transformer blocks for the inner resolution blocks so checkpointing those can help quite a bit

Reply:
Makes sense. 

From the [documentation](https://pytorch.org/docs/stable/checkpoint.html):

> Checkpointing is implemented by rerunning a forward-pass segment for each checkpointed segment during backward. This can cause persistent states like the RNG state to be advanced than they would without checkpointing. By default, checkpointing includes logic to juggle the RNG state such that checkpointed passes making use of RNG (through dropout for example) have deterministic output as compared to non-checkpointed passes. The logic to stash and restore RNG states can incur a moderate performance hit depending on the runtime of checkpointed operations. If deterministic output compared to non-checkpointed passes is not required, supply preserve_rng_state=False to checkpoint or checkpoint_sequential to omit stashing and restoring the RNG state during each checkpoint.

Do we need the comparison part between deterministic output and non-checkpointed output?  If not, we can also disable `preserve_rng_state` no?

Reply:
think that's ok that's only relevant for customized gradient backprop workflow IMO